Replace setup.py with pyproject.toml and implement optional dependencies

EwoutH opened the pull request:
This PR does two things:
 - It migrates the project install commands and metadata from `setup.py` to `pyproject.toml`. It largely follows the [convention from setuptools](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html).
 - It moves the dependencies from `requirements.txt` to the `pyproject.toml` file, and thereby also defines a few sets of optional depenencies. Note that many dependencies that were defined in the `requirements.txt` are now optional.

This PR succeeds #151 and closes #130.

**To-do:**

 - Migrate to `pyproject.toml`:
   - [x] Project and Metadata 
   - [x] Dynamic version
   - [x] Dependencies and optional dependencies
   - [x] Version 
   - [x] Packaging
 - [x] Update CI to use appropriate install commands for each job
 - [x] Update installation docs
 - [x] Implement `"all"` extras which installs everything
 - [x] Test installing `"all"` extra on Linux and macOS (does Pythonnet error?)

The installation.rst docs are also updated to reflect the new options for installing extras.

EwoutH commented:
@quaquel One thing still needs to be implemented, and I need a bit of help from you on that. That is the packaing, or what files/directories to include in the package. Currently there is [quite a large script](https://github.com/quaquel/EMAworkbench/blob/f11baaa47af4d041ac080d1ab96fd738a2b46190/setup.py#L27-L54) for, but I'm sure we could simplify that.

Could you tell me which folders and files need to be exactly included in the package? Then I will take an initial jab at implementing that in the pyproject.toml.

EwoutH commented:
With a bit of reverse engineering I figured it out. :)

The reverse-wildcard trick is brand new and not even documented in Setuptools yet (I found it in a PR, and requested it to be documented in the [Data Files Support](https://setuptools.pypa.io/en/latest/userguide/datafiles.html) docs).

EwoutH commented:
@quaquel Ready for review! Please also check the [updated installation instructions](https://github.com/quaquel/EMAworkbench/blob/pyproject/docs/source/installation.rst), and if the right dependencies are assigned to each of the extras.

quaquel commented:
You might have to walk me through all these changes in person. A couple of things
1. In the past, I had bad experiences with the automated installation of dependencies. I have seen in brake quite a few anaconda installations. This might be from before `pyproject.toml` so it might no longer be a problem
2.  I am not sure whether we want to break down the dependencies to the degree you have done here. Yes, keeping a light installation can be useful in some cases, but it also creates a possible problem for users that might not fully understand all the configuration options that they have. Again, we can discuss the exact options in person.
3. I would not call `["pysd"]` `vensim` as is now done in the [updated installation instructions](https://github.com/quaquel/EMAworkbench/blob/a3747a235da7eaad78e54c44997c5e1dc3b27450/docs/source/installation.rst). Vensim itself only requires the DSS version and is really a different connector from pysd. I suggest just calling `["pysd"]` `pysd`.

EwoutH commented:
Thanks for reviewing! I will look into your points, let's walk through them Monday.

Can we do 16:00 by the way? I won't be able to make 15:30.

EwoutH commented on pyproject.toml:
```diff
@@ -0,0 +1,57 @@
+"ema_workbench.examples.models" = ["**"]
+
+[project.optional-dependencies]
+recommended = ["ema_workbench[jupyter,dev]"]
```
Add graph

EwoutH commented:
@quaquel Turns out both JPype and pythonnet crash on other OSes than Windows.

In the latest commit:
- Added `graph` to recommended extra
- Removed the connector dependencies from `all` extra
- Updated installation.rst docs

On my side it's ready, if you agree, I will clean it up and merge!

EwoutH commented:
@quaquel If you approve I can cleanup and merge, and modify #170 to work with the new setup.

quaquel commented:
looks good so go ahead
